packages/shared/lib/services/connection.service.ts

```diff
@@ -327,6 +329,68 @@ class ConnectionService {
         return [{ connection: connection[0]!, operation: 'creation' }];
     }
 
+    public async upsertBillConnection({
+        connectionId,
+        providerConfigKey,
+        credentials,
+        connectionConfig,
+        metadata,
+        config,
+        environment,
+        account
+    }: {
+        connectionId: string;
+        providerConfigKey: string;
+        credentials: BillCredentials;
+        connectionConfig?: ConnectionConfig;
+        config: ProviderConfig;
+        metadata?: Metadata | null;
+        environment: DBEnvironment;
+        account: DBTeam;
+    }): Promise<ConnectionUpsertResponse[]> {
+        const storedConnection = await this.checkIfConnectionExists(connectionId, providerConfigKey, environment.id);
+
+        if (storedConnection) {
+            const encryptedConnection = encryptionManager.encryptConnection({
+                connection_id: connectionId,
+                config_id: config.id as number,
+                provider_config_key: providerConfigKey,
+                credentials,
+                connection_config: connectionConfig || storedConnection.connection_config,
+                environment_id: environment.id,
+                metadata: metadata || storedConnection.metadata || null
+            });
+            (encryptedConnection as Connection).updated_at = new Date();
+            const connection = await db.knex
+                .from<StoredConnection>(`_nango_connections`)
+                .where({ id: storedConnection.id!, deleted: false })
+                .update(encryptedConnection)
+                .returning('*');
+
+            void analytics.track(AnalyticsTypes.BILL_CONNECTION_INSERTED, account.id, { provider: config.provider });
+
+            return [{ connection: connection[0]!, operation: 'override' }];
+        }
+        const connection = await db.knex
+            .from<StoredConnection>(`_nango_connections`)
+            .insert(
+                encryptionManager.encryptConnection({
+                    connection_id: connectionId,
+                    provider_config_key: providerConfigKey,
+                    config_id: config.id as number,
+                    credentials,
+                    metadata: metadata || null,
+                    connection_config: connectionConfig || {},
+                    environment_id: environment.id
+                })
+            )
+            .returning('*');
```

Comment:
can we not do an upsert with `insert().onConflict()`?

Reply:
Yes this is possible, since we have constraints on the fields `provider_config_key`, `connection_id`, and `environment_id` in the connections table. I think we should also refactor the other `upsert` functions.